*: craft rangefeed abstraction, support cluster settings in tenants

ajwerner opened the pull request:
This PR is a reworking of @dt's excellent effort in #57926
to pick up rangefeeds for tenant cluster settings after sprucing up the libraries around
them. @tbg and @nvanbenschoten I'm tagging you accordingly given that you reviewed
that PR.

### First two commits (`kvclient/rangefeed` package and adoption in `catalog/lease`)

The first commit in this PR abstract out some rangefeed client logic around
retries to make them more generally usable in other contexts. The second
commit adopts this package in `sql/catalog/lease`. These two commits can
(and probably should) be merged separately and exist in
https://github.com/cockroachdb/cockroach/pull/58361.

### Next two commits (`server/settingswatcher` package and adoption in `server`)

These commits introduce code to update the settings using the above rangefeed
package and then hooks it up to the sql server.

### Next three commits

Remove the restrictions from tenants updating settings, mark which settings can be
updated, generate docs. 

These are lifted straight from @dt in #57926.

Maybe closes #56623.

dt committed:
sql: remove limitations on tenants setting cluster settings

This change also removes the limits in the SET CLUSTER SETTING
code path that prevented tenants writing to their settings table
before those changes would have had any effect.

Release note: none.

dt committed:
settings: add SystemOnly flag

This change adds a flag to indicate a setting is only applicable to the
system tenant, such as those that control storage layer behaviors, and
thus should not be set by guest tenants (where it would be ignored).

Release note: none.

dt committed:
cli/gen,Makefile: generate list of tenant-ok settings

This adds a generated listing of the settings available to tenants, i.e.
settings which are not system-only. This generated listing is checked in
so that reviewers will notice if a new setting is there that should not
be presented to tenants.

Release note: none.

Commit:
kvclient/rangefeed: introduce package for using rangefeeds

This commit introduces a new package underneath kvclient to simplify
using rangefeeds. Namely, it provides sane retry and shutdown logic as well
as support for performing an initial data scan. It also hides the nasty
unwrapping of the DistSender as performed elsewhere.

A longer-term vision here would be to support multiple spans and utilize this
layer in the `kvfeed` of `changefeedccl`.

Release note: None

Commit:
server,lease: adopt new rangefeed package

Release note: None

Commit:
server/settingswatcher: introduce a package to watch settings

This commit only barely touches existing code by extracting the logic to
decode rows from the settings table.

Release note: None

Commit:
server: hook up settingswatcher to sql server

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58362)
<!-- Reviewable:end -->

ajwerner commented:
This still breaks one test which checks for an error starting up the tenant with the wrong codec. There are some obvious fixes to work around that test but it reveals an important problem here which is that not all errors are forever retryable. Namely the tenant mismatch and a gc threshold error. I'd love to talk about how to pursue solutions there before trying something.

tbg reviewed (commented):
Nice! Thanks for putting this together. One big thing that I'm not 100% sure was on your radar is that this needs to be a "safe" backport to 20.2 if it is going to benefit CC-Serverless (as it should). Perhaps that commit you added to replace the legacy code is sufficiently isolated, but it would be easier if that commit were to be removed from this PR and merged separately to keep the backport most straightforward (your call how you go about this eventually though).

I left a ton of comments but none of them substantial, when they are addressed this is good to go as far as I'm concerned. BTW, I did not re-review David's commits; if anything in there wants another look please let me know.

@andy-kimball can you chime in on this thread here: https://reviewable.io/reviews/cockroachdb/cockroach/57926#-MOf6JkX7xTgwjcFUck5 - Essentially, do we want an allowlisting or blocklisting approach for settings? Nathan and I favor allowlisting (i.e. if you don't mark a setting as accessible to tenants, then it is not - seeing how 99% of settings currently make no sense for tenants and I worry about giving tenants ways to shoot themselves in the foot by accident) whereas David argued for a blocklist (link above).

> Reviewed 10 of 10 files at r1, 11 of 11 files at r2, 8 of 8 files at r3, 1 of 1 files at r4, 4 of 4 files at r5, 12 of 12 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-:-MQBuEOmDHxiSsYgI0Sc:bgf0x3c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBj2zo6p-79moaB9lj:-MQBj2zo6p-79moaB9lk:b-mievyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L27)):*
> ```Go
> 	retryOptions      retry.Options
> 	onInitialScanDone OnInitialScanDone
> 	withInitialScan   bool
> ```

this is just `onInitialScanDone != nil`? Might want to make it a method instead or add a comment why that won't work.
___
*[pkg/kv/kvclient/rangefeed/config.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBjMBM9TlvwqWqIMMu:-MQBjMBM9TlvwqWqIMMv:b-6dxu2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L56)):*
> ```Go
> func WithDiff() Option { return withDiff }
> 
> var withDiff = optionFunc(func(c *config) { c.withDiff = true })
> ```

nit: I personally prefer

```go
func WithDiff Option {
  return optionFunc(func(c *config) { c.withDiff = true })
}
```

because it avoids having to look around for the actual definition and is more consistent with the rest of the defs.
___
*[pkg/kv/kvclient/rangefeed/config.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBk1l5753Ql9dWgh2U:-MQBk1l5753Ql9dWgh2V:b-qga5ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L65)):*
> ```Go
> }
> 
> var defaultConfig = config{}
> ```

why this dangling var? You can just `*c = config{}` below and keep things local.
___
*[pkg/kv/kvclient/rangefeed/db_adapter.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBkADy-koVOU96F7An:-MQBkADy-koVOU96F7Ao:beh9tbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter.go#L23)):*
> ```Go
> )
> 
> // dbAdapter is an implementation of db a *kv.DB.
> ```

Sentence is garbled and also I feel it could tell me more about what this thing is. I guess it wraps a `*kv.DB` into something that allows me to run range feeds and scans?
___
*[pkg/kv/kvclient/rangefeed/db_adapter.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBlPjP8GMJwvWARzoc:-MQBlPjP8GMJwvWARzod:b-4tgzvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter.go#L43)):*
> ```Go
> 		targetScanBytes: defaultTargetScanBytes,
> 	}
> 	{
> ```

It's a shame that this is all necessary, but there's no good way around it without invasive refactors. (I was thinking to add a `DistSender()` method to `*kv.DB`, but that doesn't solve the problem - we still need to assert on the wrapped sender there because in tests it's definitely not a full dist sender). Could be fixed but let's not right now.
___
*[pkg/kv/kvclient/rangefeed/db_adapter.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBky_m5bHm8ymuqxST:-MQBky_m5bHm8ymuqxSU:b-ao1mcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter.go#L56)):*
> ```Go
> 		dbClient.distSender = distSender
> 	}
> 	return &dbClient, nil
> ```

nit: set things up so that you can return `&dbAdapter{ ... }` with all fields filled in. This pattern where we set up a partial struct and then fill in some pieces should be a last resort imo (error prone), and don't think you need it here.
___
*[pkg/kv/kvclient/rangefeed/db_adapter_external_test.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBnf2tEXYT-undQlaY:-MQBnf2tEXYT-undQlaZ:b51f8gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go#L78)):*
> ```Go
> 
> 	// Ensure that pagination doesn't break anything.
> 	cli.SetTargetScanBytes(1)
> ```

super nit: cli is also a prominent package name, maybe avoid it here.
___
*[pkg/kv/kvclient/rangefeed/doc.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBnuk63RpUYKzTqyEU:-MQBnuk78q5XwwtwbvI-:bin0g5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/doc.go#L19)):*
> ```Go
> package rangefeed
> 
> // TODO(ajwerner): Rework this logic to encapsulate the multi-span logic in
> ```

It would be useful to write out what the limitations of the current code are.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBoOOj4qCmPn6IvBZT:-MQBoOOj4qCmPn6IvBZU:b548q9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L125)):*
> ```Go
> }
> 
> // Close closes the RangeFeed and waits for it to shut down.
> ```

//
// Close is idempotent.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBoTtl47JLT9WEdeSA:-MQBoTtl47JLT9WEdeSB:b-6229oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L136)):*
> ```Go
> // failures or general unavailability. We'll reset the retrier if the
> // rangefeed runs for longer than the resetThreshold.
> const resetThreshold = 30 * time.Second
> ```

I saw this elsewhere too, but didn't understand why. Mind adding a comment?
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBode82mS1jq1w6y0r:-MQBode82mS1jq1w6y0s:ba3i4ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L139)):*
> The context is hooked up to the stopper's quiescence.

what does this mean? Is that a statement about the incoming context (in which case it's maybe more misleading than anything and could be removed).
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBpyFW9srWjgjUXmif:-MQBpyFW9srWjgjUXmig:b3lvn0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L170)):*
> ```Go
> 		start := timeutil.Now()
> 
> 		// Note that the below channel send is safe because processEvents will
> ```

is safe -> will not block forever
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBqvRWB48tRJh9hZml:-MQBpdcsC7yzttsTdC9s:b-agzdgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L171)):*
> ```Go
> 
> 		// Note that the below channel send is safe because processEvents will
> 		// wait for the worker to send. RunWorker is safe here because this
> ```

because processEvents is guaranteed to consume the error before returning.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBpRvXBGRNZLyNqtig:-MQBpRvXBGRNZLyNqtih:bqfvynu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L172)):*
> ```Go
> 		// Note that the below channel send is safe because processEvents will
> 		// wait for the worker to send. RunWorker is safe here because this
> 		// goroutine is run as an AsyncTask.
> ```

Any particular reason not to use `RunAsyncTask`? The Worker/Task API is crap and it seems that they're interchangeable here. It would be nice to make do without assumptions on the caller which tend to rot more easily than local constraints.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBrP4C0lG5je7_x5-E:-MQBrP4D7MXiTdHnRmqz:b-dy9g1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L191)):*
> ```Go
> 		if ranFor > resetThreshold {
> 			i = 1
> 			r.Reset()
> ```

This might also be a good place for a comment about the resetting.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBs3ir6RLh-JXo7k4t:-MQBs3ir6RLh-JXo7k4u:bhh9rt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L199)):*
> ```Go
> // requested. It will retry in the face of errors and will only return upon
> // context cancellation.
> func (f *RangeFeed) maybeRunInitialScan(
> ```

nit: move processEvents above this method so that it is closer to its caller.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBr_vyESDaf8PeP3hC:-MQBr_vyESDaf8PeP3hD:b-hwgi3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L203)):*
> ```Go
> ) (done bool) {
> 	if !f.withInitialScan {
> 		return ctx.Err() != nil // done
> ```

I would've expected `true` if we're not even requesting an initial scan. Why isn't that the case?
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBrnloAB3VeJJno09h:-MQBrnloAB3VeJJno09i:btwps0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L222)):*
> ```Go
> 
> 		// It's something of a bummer that we must allocate a new value for each
> 		// of these but the contract doesn't indicate that the value cannot be
> ```

s/cannot/can/?
___
*[pkg/kv/kvclient/rangefeed/rangefeed_external_test.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBsP0S3-Hb35x1l06g:-MQBsP0S3-Hb35x1l06h:bsmioz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go#L47)):*
> ```Go
> 	require.NoError(t, db.Put(ctx, mkKey("c"), 3))
> 
> 	sp := roachpb.Span{
> ```

Will it add anything to have this span multiple ranges?
___
*[pkg/kv/kvclient/rangefeed/rangefeed_external_test.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBsFEtDDXV73-m_9QW:-MQBsFEtDDXV73-m_9QX:b-7jyi4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go#L62)):*
> ```Go
> 	initialScanDone := make(chan struct{})
> 	r, err := f.RangeFeed(ctx, "test", sp, afterB, func(ctx context.Context, value *roachpb.RangeFeedValue) {
> 		fmt.Println(value)
> ```

t.Log
___
*[pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBs__S6mZqGgAllH39:-MQBs__T3ClZu_lyQmMZ:bkj7evs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go#L339)):*
> ```Go
> 	}
> 	restartingRE := regexp.MustCompile("restarting rangefeed.*after.*")
> 	log.Intercept(ctx, func(entry logpb.Entry) {
> ```

not a fan of this - this will also prevent logs from going to the files. Can you add a hook instead that you fire when you restart the rangefeed instead?
___
*[pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go, line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBstJHCf2kaaHaDZri:-MQBstJHCf2kaaHaDZrj:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go#L366)):*
> ```Go
> 		return nil
> 	})
> 	seen.Lock()
> ```

?
___
*[pkg/sql/catalog/lease/lease.go, line 1799 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBt6hp1hBBqhyzkAjP:-MQBt6hp1hBBqhyzkAjQ:b-q6uz4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad154f19b1abd1bc0ecb7dfd58047bd70a873a/pkg/sql/catalog/lease/lease.go#L1799)):*
> ```Go
> 		log.Infof(ctx, "using rangefeeds for lease manager updates")
> 	}
> 	distSender := db.NonTransactionalSender().(*kv.CrossRangeTxnWrapperSender).Wrapped().(*kvcoord.DistSender)
> ```

Are you comfortable with these changes getting backported? Since we are doing this work to enable CC, it will need a backport. Just keep that in mind - my intuition is that we should avoid touching old production code and use the new code only for tenants where it's needed. Then, on master only, we make the changes you are making here where the new package is used by existing system tenant code.


<!-- Sent from Reviewable.io -->

tbg commented:
@andy-kimball there was more discussion about allow/blocklist on the original PR, https://github.com/cockroachdb/cockroach/pull/57926

nvb reviewed (commented):
I agree with Tobi, this is a big improvement! Thanks for putting it together.

> Reviewed 10 of 10 files at r1, 11 of 11 files at r2, 8 of 8 files at r3, 1 of 1 files at r4, 4 of 4 files at r5, 12 of 12 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-:-MQDwCZO8DSGSFp_INse:b-ba3mqa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @tbg)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBj2zo6p-79moaB9lj:-MQDCv3d2wZnIR854V9p:bog5ino) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L27)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this is just `onInitialScanDone != nil`? Might want to make it a method instead or add a comment why that won't work.
</blockquote></details>

I think the intention is to allow `.WithInitialScan(nil)`.
___
*[pkg/kv/kvclient/rangefeed/config.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBjMBM9TlvwqWqIMMu:-MQDDZjzDqjkDEtrimx4:buo4lt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L56)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: I personally prefer

```go
func WithDiff Option {
  return optionFunc(func(c *config) { c.withDiff = true })
}
```

because it avoids having to look around for the actual definition and is more consistent with the rest of the defs.
</blockquote></details>

But that's more expensive than what's here, as it results in a pair of allocations per `WithDiff` call, instead of statically performing both.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDG5lk3XUoc6ewbkuF:-MQDG5lk3XUoc6ewbkuG:b-1on2v3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L134)):*
> ```
> retrier
> ```

typo? `frontier`?
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDGX5k8yv6mQaWyv1h:-MQDGX5k8yv6mQaWyv1i:b-vi0uqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L198)):*
> ```Go
> // maybeRunInitialScan will attempt to perform an initial data scan if one was
> // requested. It will retry in the face of errors and will only return upon
> // context cancellation.
> ```

Or when the scan completes.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBs3ir6RLh-JXo7k4t:-MQDGbfBDMkPlLuBSCc7:b-96gcjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L199)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: move processEvents above this method so that it is closer to its caller.
</blockquote></details>

I'm actually +1 on this order. It's the order that the methods are called from `run`.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBr_vyESDaf8PeP3hC:-MQDGpRm9d8po8y9bYek:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L203)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I would've expected `true` if we're not even requesting an initial scan. Why isn't that the case?
</blockquote></details>

+1
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDH-ao61AxilkjnSMR:-MQDH-ao61AxilkjnSMS:b-s8dayg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L215)):*
> ```Go
> 		v.Value.Timestamp = f.initialTimestamp
> 
> 		// Supply the value from the scan as also the previous value to avoid
> ```

Do we do this in the changefeed processor as well?
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDHtWD5K4-X63VowFQ:-MQDHtWD5K4-X63VowFR:b-z63q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L266)):*
> ```Go
> 			// Ensure that the RangeFeed goroutine stops.
> 			<-errCh
> 			return nil
> ```

Should we be returning `ctx.Err()`?
___
*[pkg/server/settingswatcher/settings_watcher.go, line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDtwdX3dYxYm6KdGVE:-MQDtwdX3dYxYm6KdGVF:bjwe2ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/39d6cfe9fd10d9a09ca855d109ec42ac045a9086/pkg/server/settingswatcher/settings_watcher.go#L44)):*
> ```Go
> 	clock *hlc.Clock,
> 	codec keys.SQLCodec,
> 	settings *cluster.Settings,
> ```

It wasn't as clear to me as it probably should have been that the purpose of the `SettingsWatcher` was to update this provided Settings object. Can we make that more clear through a combination of improving the commentary and maybe renaming this to `settingsToUpdate`?
___
*[pkg/server/settingswatcher/settings_watcher.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDuMVK08eeJnM-PQ2m:-MQDuMVK08eeJnM-PQ2n:b-ks57x4) ([raw file](https://github.com/cockroachdb/cockroach/blob/39d6cfe9fd10d9a09ca855d109ec42ac045a9086/pkg/server/settingswatcher/settings_watcher.go#L70)):*
> ```Go
> 	u := s.settings.MakeUpdater()
> 	initialScanDone := make(chan struct{})
> 	if _, err := s.f.RangeFeed(ctx, "settings", settingsTableSpan, now, func(
> ```

Do we want to leak the RangeFeed or should we provide a way for a client to stop the SettingsWatcher (and underlying RangeFeed)?
___
*[pkg/sql/set_cluster_setting.go, line 98 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDw1Xo-e2pWMx6McCl:-MQDw1Xo-e2pWMx6McCm:bx36ztl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4fbacec442badab5fd69844dada4379e396afa1/pkg/sql/set_cluster_setting.go#L98)):*
> ```Go
> 	if setting.SystemOnly() && !p.execCfg.Codec.ForSystemTenant() {
> 		return nil, pgerror.Newf(pgcode.InsufficientPrivilege,
> 			"setting %s is only setable in the system tenant", name)
> ```

"settable"
___
*[pkg/sql/catalog/lease/lease.go, line 1833 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDJmpj7lObx5b1g-vh:-MQDJmpj7lObx5b1g-vi:b-bnzgb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/197e527070519196c9896fe6040d2a3b25d08dde/pkg/sql/catalog/lease/lease.go#L1833)):*
> ```Go
> 		}
> 	}
> 	// Ignore errors here because they indicate that the server is shutting down.
> ```

Is this part of the function's contract. Should it be?


<!-- Sent from Reviewable.io -->

knz commented:
> One big thing that I'm not 100% sure was on your radar is that this needs to be a "safe" backport to 20.2 if it is going to benefit CC-Serverless (as it should). 

That's not so clear that it should. As described in the [linked review thread](https://reviewable.io/reviews/cockroachdb/cockroach/57926#-MOf6JkX7xTgwjcFUck5) the consensus seems to be that CC free tier running on v20.2 will not offer cluster settings to tenants, while shared tier running on v21.1 will.

We're going to pick up this conversation in the following new issue: #58445

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBjMBM9TlvwqWqIMMu:-MQHiTaJBgBpxqcj72gs:bi1ubor) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L56)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

But that's more expensive than what's here, as it results in a pair of allocations per `WithDiff` call, instead of statically performing both.
</blockquote></details>

That might be true ([is it actually though?](https://play.golang.org/p/w-jJ3tsDAw0)) but I think the nit still stands, as this is not code where this matters.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBs3ir6RLh-JXo7k4t:-MQHkYfB4R0EVD3uEDf1:bogtw6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L199)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm actually +1 on this order. It's the order that the methods are called from `run`.
</blockquote></details>

Oh, true. Carry on then.


<!-- Sent from Reviewable.io -->

tbg commented:
You've probably had reasons to not touch this in the interim, but just on the off chance that it just slipped the radar, friendly ping.

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBj2zo6p-79moaB9lj:-MQCVBi3EI-SNV6M0BkO:b-lk23s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L27)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this is just `onInitialScanDone != nil`? Might want to make it a method instead or add a comment why that won't work.
</blockquote></details>

I toyed with the interfaces here a lot. I'm not totally in love with where I ended up for what it's worth. I played with something that takes an `interface{ Init(...); Row(...); Done(...) }` for the initial scan. At that point it didn't really feel worth integrating into this interface. Perhaps it's not worth integrating but then I was annoyed at the need to write another retry loop.
___
*[pkg/kv/kvclient/rangefeed/config.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBjMBM9TlvwqWqIMMu:-MQCUtgU4vU5Ozz0yri3:b-rdmxir) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L56)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

That might be true ([is it actually though?](https://play.golang.org/p/w-jJ3tsDAw0)) but I think the nit still stands, as this is not code where this matters.
</blockquote></details>

over-optimizing an allocation in an allocation-prone setting. Removed.

However, I don't understand your example, it definitely is:
https://play.golang.org/p/9B0mwnXlCdF
___
*[pkg/kv/kvclient/rangefeed/config.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBk1l5753Ql9dWgh2U:-MQCV6s7Ag6FE6vvpTMY:b-wvbd48) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/config.go#L65)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

why this dangling var? You can just `*c = config{}` below and keep things local.
</blockquote></details>

Sure, was thinking there might be some non-zero value defaults.  Done.
___
*[pkg/kv/kvclient/rangefeed/db_adapter.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBkADy-koVOU96F7An:-MTaT3XU23aHNsKGIu2A:by119wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter.go#L23)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Sentence is garbled and also I feel it could tell me more about what this thing is. I guess it wraps a `*kv.DB` into something that allows me to run range feeds and scans?
</blockquote></details>

I'd rather that the reader go read the `kvDB` interface. I've made it a hair clearer. 
___
*[pkg/kv/kvclient/rangefeed/db_adapter.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBky_m5bHm8ymuqxST:-MTaTNWr3sLAllUjx0Oo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter.go#L56)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: set things up so that you can return `&dbAdapter{ ... }` with all fields filled in. This pattern where we set up a partial struct and then fill in some pieces should be a last resort imo (error prone), and don't think you need it here.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/db_adapter_external_test.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBnf2tEXYT-undQlaY:-MTaTbXs4tNuxgbaZtYz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go#L78)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

super nit: cli is also a prominent package name, maybe avoid it here.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBoOOj4qCmPn6IvBZT:-MTaTqK35F-Plm_4ME7y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L125)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

//
// Close is idempotent.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDG5lk3XUoc6ewbkuF:-MTaTrf1AgV_rQ8AxMZS:b6fa8du) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L134)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> retrier
> ```

typo? `frontier`?
</blockquote></details>

Not a typo. Reworded for clarity. 
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBoTtl47JLT9WEdeSA:-MTaUGR14djmi-DO45PX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L136)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I saw this elsewhere too, but didn't understand why. Mind adding a comment?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBode82mS1jq1w6y0r:-MTaUMHrCi5w7LgijRez:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L139)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> The context is hooked up to the stopper's quiescence.

what does this mean? Is that a statement about the incoming context (in which case it's maybe more misleading than anything and could be removed).

</blockquote></details>

Removed.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBqvRWB48tRJh9hZml:-MTaUacYD8i1_ajSx9F_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L171)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

because processEvents is guaranteed to consume the error before returning.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBpRvXBGRNZLyNqtig:-MTaUbPU8TiWOQOoKEZJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L172)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Any particular reason not to use `RunAsyncTask`? The Worker/Task API is crap and it seems that they're interchangeable here. It would be nice to make do without assumptions on the caller which tend to rot more easily than local constraints.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBrP4C0lG5je7_x5-E:-MTaUtDK-tr5UeewWGxi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L191)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This might also be a good place for a comment about the resetting.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBr_vyESDaf8PeP3hC:-MTaV0zpA2XThFZUprpt:b20y2sh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L203)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1
</blockquote></details>

I had a test that wanted to make sure we didn't kick off the rangefeed if the context is canceled prior to the `Rangefeed` call. I've reworked the context checking to make this true elsewhere. 
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDH-ao61AxilkjnSMR:-MTaW7XzDFhXq8aQSqnO:b-7xt2jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L215)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we do this in the changefeed processor as well?
</blockquote></details>

Yes. 
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBrnloAB3VeJJno09h:-MTaWDti70UsIj0_wwDm:bcz206) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L222)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

s/cannot/can/?
</blockquote></details>

I think it's right as stated. The contract doesn't say anything about ownership so we take the pessimistic view. 
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDHtWD5K4-X63VowFQ:-MTaWmgN-eC5hXziPPnw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed.go#L266)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we be returning `ctx.Err()`?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed_external_test.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBsP0S3-Hb35x1l06g:-MTaWyanC2i93WNX18Uu:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go#L47)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Will it add anything to have this span multiple ranges?
</blockquote></details>

Done. 
___
*[pkg/kv/kvclient/rangefeed/rangefeed_external_test.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBsFEtDDXV73-m_9QW:-MQCWDWjE1X8F2vvl8JH:bul0v5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go#L62)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

t.Log
</blockquote></details>

detritus, removed
___
*[pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBs__S6mZqGgAllH39:-MTaeSKT8ACfkW56kgWo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go#L339)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

not a fan of this - this will also prevent logs from going to the files. Can you add a hook instead that you fire when you restart the rangefeed instead?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go, line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBstJHCf2kaaHaDZri:-MQCWKXl4bTZ2Z9gvdBB:b-fkiw5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go#L366)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

?
</blockquote></details>

detritus, removed. 
___
*[pkg/server/settingswatcher/settings_watcher.go, line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDtwdX3dYxYm6KdGVE:-MTa_nnjCFHt6M5vxDYR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/39d6cfe9fd10d9a09ca855d109ec42ac045a9086/pkg/server/settingswatcher/settings_watcher.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It wasn't as clear to me as it probably should have been that the purpose of the `SettingsWatcher` was to update this provided Settings object. Can we make that more clear through a combination of improving the commentary and maybe renaming this to `settingsToUpdate`?
</blockquote></details>

Done.
___
*[pkg/server/settingswatcher/settings_watcher.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDuMVK08eeJnM-PQ2m:-MTa_XNkDZ1Yf-fe-j8Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/39d6cfe9fd10d9a09ca855d109ec42ac045a9086/pkg/server/settingswatcher/settings_watcher.go#L70)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want to leak the RangeFeed or should we provide a way for a client to stop the SettingsWatcher (and underlying RangeFeed)?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/lease.go, line 1799 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBt6hp1hBBqhyzkAjP:-MQCU_dl3fqOwqh6BkyC:b-2b0dxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad154f19b1abd1bc0ecb7dfd58047bd70a873a/pkg/sql/catalog/lease/lease.go#L1799)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Are you comfortable with these changes getting backported? Since we are doing this work to enable CC, it will need a backport. Just keep that in mind - my intuition is that we should avoid touching old production code and use the new code only for tenants where it's needed. Then, on master only, we make the changes you are making here where the new package is used by existing system tenant code.
</blockquote></details>

We're not trying to backport this. 
___
*[pkg/sql/catalog/lease/lease.go, line 1833 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQDJmpj7lObx5b1g-vh:-MTa_MSvBFK1Bngmn3zF:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/197e527070519196c9896fe6040d2a3b25d08dde/pkg/sql/catalog/lease/lease.go#L1833)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this part of the function's contract. Should it be?
</blockquote></details>

Added.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
___
*[pkg/sql/catalog/lease/lease.go, line 1799 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBt6hp1hBBqhyzkAjP:-MTb8sQu2SNq1KpzKgGH:b-e1v8ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad154f19b1abd1bc0ecb7dfd58047bd70a873a/pkg/sql/catalog/lease/lease.go#L1799)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We're not trying to backport this. 
</blockquote></details>

Andrew's correct: this does not need to be back-ported, since we mainly need this for supporting upcoming features that are part of 21.1. Of course, that assumes we can upgrade from 20.2 to 21.1, which is not yet possible, but that's a different discussion.


<!-- Sent from Reviewable.io -->

ajwerner commented:
This should be RFAL. Sorry I let it rot for so long. Working on getting it in and doing something on top to allow us to get tenants upgraded.

andy-kimball reviewed (commented):
So if this is necessary for getting tenants upgraded from 20.2 => 21.1, is it also safe to back-port to 20.2?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-:-MTdeEhu5o96FbmH5d96:bkp7ic5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

ajwerner commented:
> So if this is necessary for getting tenants upgraded from 20.2 => 21.1, is it also safe to back-port to 20.2?
> 
> > *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-:-MTdeEhu5o96FbmH5d96:bkp7ic5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @nvanbenschoten, and @tbg)
> 
> 
> 
> <!-- Sent from Reviewable.io -->
> 

No, but I don't think that matters. This change is compatible with 20.2. What we need is to make sure that when we boot 21.1 sql servers they can activate the 21.1 cluster version.

Commit:
stop/stopper: use sync.Map for contexts to allow canceling contexts in close

Before this commit, if a Closer called cancel on a context which was attached
to the stopper, it would result in a deadlock as Close is called under the
same mutex that that the cancel method was trying to lock. This exchanges that
map for an atomic map.

Release note: None

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 44 files at r8, 6 of 10 files at r9, 11 of 15 files at r10, 9 of 9 files at r11, 1 of 1 files at r12, 1 of 4 files at r13, 11 of 12 files at r14, 4 of 4 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-:-MTi15BkEX7grPAkIep8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @tbg)

___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 114 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MThzC39DYz3Oumr9WV-:-MThzC39DYz3Oumr9WV0:bizjzj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b746df656b37bbab55d094cbe083420ce64d0fba/pkg/kv/kvclient/rangefeed/rangefeed.go#L114)):*
> ```Go
> 		client:  f.client,
> 		stopper: f.stopper,
> 		knobs:   &f.knobs,
> ```

Should we just store the TestingKnobs as a pointer on Factory?
___
*[pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go, line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MQBstJHCf2kaaHaDZri:-MTi-iP93gny-FvcVHVd:bu4yk5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7577b9f1f3e04b6739ad84df43e8fae235efa5/pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go#L366)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

detritus, removed. 
</blockquote></details>

It doesn't look like this was removed.
___
*[pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go, line 332 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58362#-MTi-cVN2SQWetUwhWYo:-MTi-cVN2SQWetUwhWYp:b-4ok724) ([raw file](https://github.com/cockroachdb/cockroach/blob/b746df656b37bbab55d094cbe083420ce64d0fba/pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go#L332)):*
> ```Go
> 	}
> 	ctx := context.Background()
> 	var seen struct {
> ```

nit: would an atomic int be simpler?


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r=nvanbenschoten
